File: java/src/main/java/io/cucumber/query/Query.java
package io.cucumber.query;

import io.cucumber.messages.Convertor;
import io.cucumber.messages.TestStepResultStatusComparator;
import io.cucumber.messages.types.Attachment;
import io.cucumber.messages.types.Examples;
import io.cucumber.messages.types.Feature;
import io.cucumber.messages.types.GherkinDocument;
import io.cucumber.messages.types.Hook;
import io.cucumber.messages.types.Location;
import io.cucumber.messages.types.Meta;
import io.cucumber.messages.types.Pickle;
import io.cucumber.messages.types.PickleStep;
import io.cucumber.messages.types.Rule;
import io.cucumber.messages.types.Scenario;
import io.cucumber.messages.types.Step;
import io.cucumber.messages.types.StepDefinition;
import io.cucumber.messages.types.Suggestion;
import io.cucumber.messages.types.TableRow;
import io.cucumber.messages.types.TestCase;
import io.cucumber.messages.types.TestCaseFinished;
import io.cucumber.messages.types.TestCaseStarted;
import io.cucumber.messages.types.TestRunFinished;
import io.cucumber.messages.types.TestRunStarted;
import io.cucumber.messages.types.TestStep;
import io.cucumber.messages.types.TestStepFinished;
import io.cucumber.messages.types.TestStepResult;
import io.cucumber.messages.types.TestStepResultStatus;
import io.cucumber.messages.types.TestStepStarted;
import io.cucumber.messages.types.Timestamp;

import java.time.Duration;
import java.util.AbstractMap.SimpleEntry;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.EnumMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.Objects;

import static java.util.Collections.emptyList;
import static java.util.Comparator.comparing;
import static java.util.Objects.requireNonNull;
import static java.util.Optional.ofNullable;
import static java.util.function.Function.identity;
import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toList;

/**
 * Given one Cucumber Message, find another.
 * <p>
 * Queries can be made while the test run is incomplete - and this will
 * naturally return incomplete results.
 *
 * @see <a href="https://github.com/cucumber/messages?tab=readme-ov-file#message-overview">Cucumber Messages - Message Overview</a>
 */
public final class Query {
    private final Repository repository;

    public Query(Repository repository) {
        this.repository = repository;
    }

    public Map<TestStepResultStatus, Long> countMostSevereTestStepResultStatus() {
        EnumMap<TestStepResultStatus, Long> results = new EnumMap<>(TestStepResultStatus.class);
        for (TestStepResultStatus value : TestStepResultStatus.values()) {
            results.put(value, 0L);
        }
        results.putAll(findAllTestCaseStarted().stream()
                .map(this::findMostSevereTestStepResultBy)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .map(TestStepResult::getStatus)
                .collect(groupingBy(identity(), LinkedHashMap::new, counting())));
        return results;
    }
    public int countTestCasesStarted() {
        return findAllTestCaseStarted().size();
    }

    public List<Pickle> findAllPickles() {
        return new ArrayList<>(repository.pickleById.values());
    }

    public List<PickleStep> findAllPickleSteps() {
        return new ArrayList<>(repository.pickleStepById.values());
    }

    public List<TestCaseStarted> findAllTestCaseStarted() {
        return repository.testCaseStartedById.values().stream()
                .filter(element -> !findTestCaseFinishedBy(element)
                        .filter(TestCaseFinished::getWillBeRetried)
                        .isPresent())
                .collect(toList());
    }

    public List<TestCaseFinished> findAllTestCaseFinished() {
        return repository.testCaseFinishedByTestCaseStartedId.values().stream()
                .filter(testCaseFinished -> !testCaseFinished.getWillBeRetried())
                .collect(toList());
    }

    public List<TestStep> findAllTestSteps() {
        return new ArrayList<>(repository.testStepById.values());
    }

    public List<TestCase> findAllTestCases() {
        return new ArrayList<>(repository.testCaseById.values());
    }

    public List<TestStepStarted> findAllTestStepStarted() {
        return repository.testStepsStartedByTestCaseStartedId.values().stream()
                .flatMap(Collection::stream)
                .collect(toList());
    }

    public List<TestStepFinished> findAllTestStepFinished() {
        return repository.testStepsFinishedByTestCaseStartedId.values().stream()
                .flatMap(Collection::stream)
                .collect(toList());
    }

    public List<Attachment> findAttachmentsBy(TestStepFinished testStepFinished) {
        requireNonNull(testStepFinished);
        return repository.attachmentsByTestCaseStartedId.getOrDefault(testStepFinished.getTestCaseStartedId(), emptyList()).stream()
                .filter(attachment -> attachment.getTestStepId()
                        .map(testStepId -> testStepFinished.getTestStepId().equals(testStepId))
                        .orElse(false))
                .collect(toList());
    }

    public Optional<Hook> findHookBy(TestStep testStep) {
        requireNonNull(testStep);
        return testStep.getHookId()
                .map(repository.hookById::get);
    }

    public Optional<Meta> findMeta() {
        return ofNullable(repository.meta);
    }

    public Optional<TestStepResult> findMostSevereTestStepResultBy(TestCaseStarted testCaseStarted) {
        requireNonNull(testCaseStarted);
        return findTestStepsFinishedBy(testCaseStarted)
                .stream()
                .map(TestStepFinished::getTestStepResult)
                .max(comparing(TestStepResult::getStatus, new TestStepResultStatusComparator()));
    }

    public Optional<TestStepResult> findMostSevereTestStepResultBy(TestCaseFinished testCaseFinished) {
        requireNonNull(testCaseFinished);
        return findTestCaseStartedBy(testCaseFinished)
                .flatMap(this::findMostSevereTestStepResultBy);
    }

    public Optional<Location> findLocationOf(Pickle pickle) {
        return findLineageBy(pickle).flatMap(lineage -> {
            if (lineage.example().isPresent()) {
                return lineage.example().map(TableRow::getLocation);
            }
            return lineage.scenario().map(Scenario::getLocation);
        });
    }

    public Optional<Pickle> findPickleBy(TestCaseStarted testCaseStarted) {
        requireNonNull(testCaseStarted);
        return findTestCaseBy(testCaseStarted)
                .flatMap(this::findPickleBy);
    }

    public Optional<Pickle> findPickleBy(TestCaseFinished testCaseFinished) {
        requireNonNull(testCaseFinished);
        return findTestCaseStartedBy(testCaseFinished)
                .flatMap(this::findPickleBy);
    }

    public Optional<Pickle> findPickleBy(TestCase testCase) {
        requireNonNull(testCase);
        return ofNullable(repository.pickleById.get(testCase.getPickleId()));
    }

    public Optional<Pickle> findPickleBy(TestStepStarted testStepStarted) {
        requireNonNull(testStepStarted);
        return findTestCaseBy(testStepStarted)
                .map(TestCase::getPickleId)
                .map(repository.pickleById::get);
    }

    public Optional<Pickle> findPickleBy(TestStepFinished testStepFinished) {
        requireNonNull(testStepFinished);
        return findTestCaseBy(testStepFinished)
                .map(TestCase::getPickleId)
                .map(repository.pickleById::get);
    }

    public Optional<PickleStep> findPickleStepBy(TestStep testStep) {
        requireNonNull(testStep);
        return testStep.getPickleStepId()
                .map(repository.pickleStepById::get);
    }

    public List<Suggestion> findSuggestionsBy(PickleStep pickleStep) {
        requireNonNull(pickleStep);
        List<Suggestion> suggestions = repository.suggestionsByPickleStepId.getOrDefault(pickleStep.getId(), emptyList());
        return new ArrayList<>(suggestions);
    }

    public List<Suggestion> findSuggestionsBy(Pickle pickle) {
        requireNonNull(pickle);
        return pickle.getSteps().stream()
                .map(this::findSuggestionsBy)
                .flatMap(Collection::stream)
                .collect(toList());
    }

    public Optional<Step> findStepBy(PickleStep pickleStep) {
        requireNonNull(pickleStep);
        String stepId = pickleStep.getAstNodeIds().get(0);
        return ofNullable(repository.stepById.get(stepId));
    }

    public List<StepDefinition> findStepDefinitionsBy(TestStep testStep) {
        requireNonNull(testStep);
        return testStep.getStepDefinitionIds().map(ids -> ids.stream()
                        .map(repository.stepDefinitionById::get)
                        .filter(Objects::nonNull)
                        .collect(toList()))
                .orElseGet(Collections::emptyList);
    }

    public Optional<StepDefinition> findUnambiguousStepDefinitionBy(TestStep testStep) {
        requireNonNull(testStep);
        return testStep.getStepDefinitionIds()
                .filter(ids -> ids.size() == 1)
                .map(ids -> repository.stepDefinitionById.get(ids.get(0)));
    }

    public Optional<TestCase> findTestCaseBy(TestCaseStarted testCaseStarted) {
        requireNonNull(testCaseStarted);
        return ofNullable(repository.testCaseById.get(testCaseStarted.getTestCaseId()));
    }

    public Optional<TestCase> findTestCaseBy(TestCaseFinished testCaseFinished) {
        requireNonNull(testCaseFinished);
        return findTestCaseStartedBy(testCaseFinished)
                .flatMap(this::findTestCaseBy);
    }

    public Optional<TestCase> findTestCaseBy(TestStepStarted testStepStarted) {
        requireNonNull(testStepStarted);
        return findTestCaseStartedBy(testStepStarted)
                .flatMap(this::findTestCaseBy);
    }

    public Optional<TestCase> findTestCaseBy(TestStepFinished testStepFinished) {
        requireNonNull(testStepFinished);
        return findTestCaseStartedBy(testStepFinished)
                .flatMap(this::findTestCaseBy);
    }

    public Optional<Duration> findTestCaseDurationBy(TestCaseStarted testCaseStarted) {
        requireNonNull(testCaseStarted);
        Timestamp started = testCaseStarted.getTimestamp();
        return findTestCaseFinishedBy(testCaseStarted)
                .map(TestCaseFinished::getTimestamp)
                .map(finished -> Duration.between(
                        Convertor.toInstant(started),
                        Convertor.toInstant(finished)
                ));
    }

    public Optional<Duration> findTestCaseDurationBy(TestCaseFinished testCaseFinished) {
        requireNonNull(testCaseFinished);
        return findTestCaseStartedBy(testCaseFinished)
                .flatMap(this::findTestCaseDurationBy);
    }

    public Optional<TestCaseStarted> findTestCaseStartedBy(TestStepStarted testStepStarted) {
        requireNonNull(testStepStarted);
        String testCaseStartedId = testStepStarted.getTestCaseStartedId();
        return ofNullable(repository.testCaseStartedById.get(testCaseStartedId));
    }

    public Optional<TestCaseStarted> findTestCaseStartedBy(TestCaseFinished testCaseFinished) {
        requireNonNull(testCaseFinished);
        String testCaseStartedId = testCaseFinished.getTestCaseStartedId();
        return ofNullable(repository.testCaseStartedById.get(testCaseStartedId));
    }

    public Optional<TestCaseStarted> findTestCaseStartedBy(TestStepFinished testStepFinished) {
        requireNonNull(testStepFinished);
        String testCaseStartedId = testStepFinished.getTestCaseStartedId();
        return ofNullable(repository.testCaseStartedById.get(testCaseStartedId));
    }

    public Optional<TestCaseFinished> findTestCaseFinishedBy(TestCaseStarted testCaseStarted) {
        requireNonNull(testCaseStarted);
        return ofNullable(repository.testCaseFinishedByTestCaseStartedId.get(testCaseStarted.getId()));
    }

    public Optional<Duration> findTestRunDuration() {
        if (repository.testRunStarted == null || repository.testRunFinished == null) {
            return Optional.empty();
        }
        Duration between = Duration.between(
                Convertor.toInstant(repository.testRunStarted.getTimestamp()),
                Convertor.toInstant(repository.testRunFinished.getTimestamp())
        );
        return Optional.of(between);
    }

    public Optional<TestRunFinished> findTestRunFinished() {
        return ofNullable(repository.testRunFinished);
    }

    public Optional<TestRunStarted> findTestRunStarted() {
        return ofNullable(repository.testRunStarted);
    }

    public Optional<TestStep> findTestStepBy(TestStepStarted testStepStarted) {
        requireNonNull(testStepStarted);
        return ofNullable(repository.testStepById.get(testStepStarted.getTestStepId()));
    }

    public Optional<TestStep> findTestStepBy(TestStepFinished testStepFinished) {
        requireNonNull(testStepFinished);
        return ofNullable(repository.testStepById.get(testStepFinished.getTestStepId()));
    }

    public List<TestStepStarted> findTestStepsStartedBy(TestCaseStarted testCaseStarted) {
        requireNonNull(testCaseStarted);
        List<TestStepStarted> testStepsStarted = repository.testStepsStartedByTestCaseStartedId.
                getOrDefault(testCaseStarted.getId(), emptyList());
        // Concurrency
        return new ArrayList<>(testStepsStarted);
    }

    public List<TestStepStarted> findTestStepsStartedBy(TestCaseFinished testCaseFinished) {
        requireNonNull(testCaseFinished);
        List<TestStepStarted> testStepsStarted = repository.testStepsStartedByTestCaseStartedId.
                getOrDefault(testCaseFinished.getTestCaseStartedId(), emptyList());
        // Concurrency
        return new ArrayList<>(testStepsStarted);
    }

    public List<TestStepFinished> findTestStepsFinishedBy(TestCaseStarted testCaseStarted) {
        requireNonNull(testCaseStarted);
        List<TestStepFinished> testStepsFinished = repository.testStepsFinishedByTestCaseStartedId.
                getOrDefault(testCaseStarted.getId(), emptyList());
        // Concurrency
        return new ArrayList<>(testStepsFinished);
    }

    public List<TestStepFinished> findTestStepsFinishedBy(TestCaseFinished testCaseFinished) {
        requireNonNull(testCaseFinished);
        return findTestCaseStartedBy(testCaseFinished)
                .map(this::findTestStepsFinishedBy)
                .orElseGet(ArrayList::new);
    }

    public List<Entry<TestStepFinished, TestStep>> findTestStepFinishedAndTestStepBy(TestCaseStarted testCaseStarted) {
        return findTestStepsFinishedBy(testCaseStarted).stream()
                .map(testStepFinished -> findTestStepBy(testStepFinished).map(testStep -> new SimpleEntry<>(testStepFinished, testStep)))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(toList());
    }

    public Optional<Lineage> findLineageBy(GherkinDocument element) {
        requireNonNull(element);
        return Optional.ofNullable(repository.lineageById.get(element.getUri()));
    }

    public Optional<Lineage> findLineageBy(Feature element) {
        requireNonNull(element);
        return Optional.ofNullable(repository.lineageById.get(element));
    }

    public Optional<Lineage> findLineageBy(Rule element) {
        requireNonNull(element);
        return Optional.ofNullable(repository.lineageById.get(element.getId()));
    }

    public Optional<Lineage> findLineageBy(Scenario element) {
        requireNonNull(element);
        return Optional.ofNullable(repository.lineageById.get(element.getId()));
    }

    public Optional<Lineage> findLineageBy(Examples element) {
        requireNonNull(element);
        return Optional.ofNullable(repository.lineageById.get(element.getId()));
    }

    public Optional<Lineage> findLineageBy(TableRow element) {
        requireNonNull(element);
        return Optional.ofNullable(repository.lineageById.get(element.getId()));
    }

    public Optional<Lineage> findLineageBy(Pickle pickle) {
        requireNonNull(pickle);
        List<String> astNodeIds = pickle.getAstNodeIds();
        String pickleAstNodeId = astNodeIds.get(astNodeIds.size() - 1);
        return Optional.ofNullable(repository.lineageById.get(pickleAstNodeId));
    }

    public Optional<Lineage> findLineageBy(TestCaseStarted testCaseStarted) {
        return findPickleBy(testCaseStarted)
                .flatMap(this::findLineageBy);
    }
}

File: dotnet/Query/Query/Query.cs
#nullable enable
using System;
using System.ComponentModel.DataAnnotations;
using System.Collections.Generic;
using System.Linq;
using Io.Cucumber.Messages.Types;
using System.Collections.Concurrent;
using Cucumber.Messages;

namespace Io.Cucumber.Query;

// Ported from io.cucumber.query.Query (Java)
public class Query
{
    private readonly Repository _repository;

    public Query(Repository repository)
    {
        _repository = repository;
    }

    // 1. countMostSevereTestStepResultStatus
    public IDictionary<TestStepResultStatus, long> CountMostSevereTestStepResultStatus()
    {
        var statusCounts = new Dictionary<TestStepResultStatus, long>();
        // Initialize with zero for each possible TestStepResultStatus
        foreach (TestStepResultStatus status in Enum.GetValues(typeof(TestStepResultStatus)))
        {
            statusCounts[status] = 0;
        }
        foreach (var testCaseStarted in FindAllTestCaseStarted())
        {
            var finishedSteps = FindTestStepsFinishedBy(testCaseStarted);
            if (finishedSteps.Count == 0)
                continue;
            // Find the most severe status (largest enum value)
            var mostSevere = finishedSteps
                .Select(f => f.TestStepResult.Status)
                .Max();
            if (statusCounts.ContainsKey(mostSevere))
                statusCounts[mostSevere]++;
            else
                statusCounts[mostSevere] = 1;
        }
        return statusCounts;
    }

    // 2. countTestCasesStarted
    public int TestCasesStartedCount => FindAllTestCaseStarted().Count;

    // 3. FindAllPickles
    public IList<Pickle> FindAllPickles() => _repository.PickleById.Values.ToList();

    // 4. FindAllPickleSteps
    public IList<PickleStep> FindAllPickleSteps() => _repository.PickleStepById.Values.ToList();

    // 5. FindAllTestCaseStarted
    public IList<TestCaseStarted> FindAllTestCaseStarted() => _repository.TestCaseStartedById.Values
        .Where(tcs => !FindTestCaseFinishedBy(tcs)?.WillBeRetried ?? true) // Exclude finished cases that will be retried
        .ToList();

    // 6. FindAllTestCaseFinished
    public IList<TestCaseFinished> FindAllTestCaseFinished()
    {
        return _repository.TestCaseFinishedByTestCaseStartedId.Values
            .Where(tcFinished => !tcFinished.WillBeRetried)
            .ToList();
    }

    // 7. FindAllTestSteps
    public IList<TestStep> FindAllTestSteps() => _repository.TestStepById.Values.ToList();

    // 8. FindAllTestCases
    public IList<TestCase> FindAllTestCases()
    {
        return _repository.TestCaseById.Values.ToList();
    }

    // 9. FindAllTestStepStarted
    public IList<TestStepStarted> FindAllTestStepStarted()
    {
        return _repository.TestStepsStartedByTestCaseStartedId.Values
            .SelectMany(list => list)
            .ToList();
    }

    // 10. FindAllTestStepFinished
    public IList<TestStepFinished> FindAllTestStepFinished()
    {
        return _repository.TestStepsFinishedByTestCaseStartedId.Values
            .SelectMany(list => list)
            .ToList();
    }

    // 11. FindAllTestRunHookStarted
    public IList<TestRunHookStarted> FindAllTestRunHookStarted()
    {
        return _repository.TestRunHookStartedById.Values.ToList();
    }

    // 12. FindAllTestRunHookFinished
    public IList<TestRunHookFinished> FindAllTestRunHookFinished()
    {
        return _repository.TestRunHookFinishedByTestRunHookStartedId.Values.ToList();
    }

    // 13. FindAttachmentsBy(TestStepFinished)
    public IList<Attachment> FindAttachmentsBy(TestStepFinished testStepFinished) =>
        _repository.AttachmentsByTestCaseStartedId.TryGetValue(testStepFinished.TestCaseStartedId, out var attachments)
            ? attachments.Where(a => a.TestStepId == testStepFinished.TestStepId).ToList()
            : new List<Attachment>();

    // 14. FindAttachmentsBy(TestRunHookFinished)
    public IList<Attachment> FindAttachmentsBy(TestRunHookFinished testRunHookFinished)
    {
        if (_repository.AttachmentsByTestRunHookStartedId.TryGetValue(testRunHookFinished.TestRunHookStartedId, out var attachments))
            return new List<Attachment>(attachments);
        return new List<Attachment>();
    }

    // 15. FindHookBy
    public Hook? FindHookBy(TestStep testStep)
    {
        if (!string.IsNullOrEmpty(testStep.HookId) && _repository.HookById.TryGetValue(testStep.HookId, out var hook))
        {
            return hook;
        }
        return null;
    }

    public Hook? FindHookBy(TestRunHookStarted testRunHookStarted)
    {
        if (!string.IsNullOrEmpty(testRunHookStarted.HookId) && _repository.HookById.TryGetValue(testRunHookStarted.HookId, out var hook))
        {
            return hook;
        }
        return null;
    }

    public Hook? FindHookBy(TestRunHookFinished testRunHookFinished)
    {
        var testRunHookStarted = FindTestRunHookStartedBy(testRunHookFinished);
        if (testRunHookStarted == null)
        {
            return null;
        }
        return FindHookBy(testRunHookStarted);
    }

    // 16. FindMeta
    public Meta? FindMeta() => _repository.Meta;

    // 17. FindMostSevereTestStepResultBy(TestCaseStarted)
    public TestStepResult? FindMostSevereTestStepResultBy(TestCaseStarted testCaseStarted)
    {
        var finishedSteps = FindTestStepsFinishedBy(testCaseStarted);
        if (finishedSteps.Count == 0)
            return null;
        // Find the TestStepFinished with the most severe status (highest enum value)
        var mostSevere = finishedSteps
            .OrderBy(f => f.TestStepResult.Status)
            .LastOrDefault();
        return mostSevere?.TestStepResult;
    }

    // 18. FindMostSevereTestStepResultBy(TestCaseFinished)
    public TestStepResult? FindMostSevereTestStepResultBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindMostSevereTestStepResultBy(testCaseStarted) : null;
    }

    // 19. FindLocationOf
    public Location? FindLocationOf(Pickle pickle)
    {
        var lineage = FindLineageBy(pickle);
        if (lineage == null)
            return null;
        if (lineage.Example != null)
            return lineage.Example.Location;
        if (lineage.Scenario != null)
            return lineage.Scenario.Location;
        return null;
    }

    // 20. FindPickleBy(TestCaseStarted)
    public Pickle? FindPickleBy(TestCaseStarted testCaseStarted)
    {
        var testCase = FindTestCaseBy(testCaseStarted);
        if (testCase != null && _repository.PickleById.TryGetValue(testCase.PickleId, out var pickle))
        {
            return pickle;
        }
        return null;
    }

    // 21. FindPickleBy(TestCaseFinished)
    public Pickle? FindPickleBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindPickleBy(testCaseStarted) : null;
    }

    // 22. FindPickleBy(TestCase)
    public Pickle? FindPickleBy(TestCase testCase)
    {
        return testCase != null && _repository.PickleById.TryGetValue(testCase.PickleId, out var pickle) ? pickle : null;
    }

    // 23. FindPickleBy(TestStepStarted)
    public Pickle? FindPickleBy(TestStepStarted testStepStarted)
    {
        var testCaseStarted = FindTestCaseStartedBy(testStepStarted);
        return testCaseStarted != null ? FindPickleBy(testCaseStarted) : null;
    }

    // 24. FindPickleBy(TestStepFinished)
    public Pickle? FindPickleBy(TestStepFinished testStepFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testStepFinished);
        return testCaseStarted != null ? FindPickleBy(testCaseStarted) : null;
    }

    // 25. FindPickleStepBy
    public PickleStep? FindPickleStepBy(TestStep testStep)
    {
        if (!string.IsNullOrEmpty(testStep.PickleStepId))
        {
            if (_repository.PickleStepById.TryGetValue(testStep.PickleStepId, out var pickleStep))
            {
                return pickleStep;
            }
        }
        return null;
    }

    // 26. FindSuggestionsBy(PickleStep)
    public IList<Suggestion> FindSuggestionsBy(PickleStep pickleStep)
    {
        if (_repository.SuggestionsByPickleStepId.TryGetValue(pickleStep.Id, out var suggestions))
            return new List<Suggestion>(suggestions);
        return new List<Suggestion>();
    }

    // 27. FindSuggestionsBy(Pickle)
    public IList<Suggestion> FindSuggestionsBy(Pickle pickle)
    {
        var result = new List<Suggestion>();
        foreach (var step in pickle.Steps)
            result.AddRange(FindSuggestionsBy(step));
        return result;
    }

    // 28. FindStepBy(PickleStep)
    public Step? FindStepBy(PickleStep pickleStep)
    {
        if (pickleStep.AstNodeIds != null && pickleStep.AstNodeIds.Count > 0)
        {
            var stepId = pickleStep.AstNodeIds[0];
            if (!string.IsNullOrEmpty(stepId) && _repository.StepById.TryGetValue(stepId, out var step))
            {
                return step;
            }
        }
        return null;
    }

    // 29. FindStepDefinitionsBy
    public IList<StepDefinition> FindStepDefinitionsBy(TestStep testStep)
    {
        if (testStep.StepDefinitionIds != null)
            return testStep.StepDefinitionIds
                .Select(id => _repository.StepDefinitionById.TryGetValue(id, out var def) ? def : null)
                .Where(def => def != null)
                .Cast<StepDefinition>()
                .ToList();
        return new List<StepDefinition>();
    }

    // 30. FindUnambiguousStepDefinitionBy
    public StepDefinition? FindUnambiguousStepDefinitionBy(TestStep testStep)
    {
        if (testStep.StepDefinitionIds != null && testStep.StepDefinitionIds.Count == 1)
            return _repository.StepDefinitionById.TryGetValue(testStep.StepDefinitionIds[0], out var def) ? def : null;
        return null;
    }

    // 31. FindTestCaseBy(TestCaseStarted)
    public TestCase? FindTestCaseBy(TestCaseStarted testCaseStarted)
    {
        if (_repository.TestCaseById.TryGetValue(testCaseStarted.TestCaseId, out var testCase))
        {
            return testCase;
        }
        return null;
    }

    // 32. FindTestCaseBy(TestCaseFinished)
    public TestCase? FindTestCaseBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindTestCaseBy(testCaseStarted) : null;
    }

    // 33. FindTestCaseBy(TestStepStarted)
    public TestCase? FindTestCaseBy(TestStepStarted testStepStarted)
    {
        var testCaseStarted = FindTestCaseStartedBy(testStepStarted);
        return testCaseStarted != null ? FindTestCaseBy(testCaseStarted) : null;
    }

    // 34. FindTestCaseBy(TestStepFinished)
    public TestCase? FindTestCaseBy(TestStepFinished testStepFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testStepFinished);
        return testCaseStarted != null ? FindTestCaseBy(testCaseStarted) : null;
    }

    // 35. FindTestCaseDurationBy(TestCaseStarted)
    public System.TimeSpan? FindTestCaseDurationBy(TestCaseStarted testCaseStarted)
    {
        var started = testCaseStarted.Timestamp;
        var finished = FindTestCaseFinishedBy(testCaseStarted)?.Timestamp;
        if (finished != null)
        {
            var startTime = Converters.ToDateTime(started);
            var finishTime = Converters.ToDateTime(finished);
            return finishTime - startTime;
        }
        return null;
    }

    // 36. FindTestCaseDurationBy(TestCaseFinished)
    public System.TimeSpan? FindTestCaseDurationBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindTestCaseDurationBy(testCaseStarted) : null;
    }

    // 37. FindTestCaseStartedBy(TestStepStarted)
    public TestCaseStarted? FindTestCaseStartedBy(TestStepStarted testStepStarted)
    {
        return _repository.TestCaseStartedById.TryGetValue(testStepStarted.TestCaseStartedId, out var tcs) ? tcs : null;
    }

    // 38. FindTestCaseStartedBy(TestCaseFinished)
    public TestCaseStarted? FindTestCaseStartedBy(TestCaseFinished testCaseFinished)
    {
        return _repository.TestCaseStartedById.TryGetValue(testCaseFinished.TestCaseStartedId, out var tcs) ? tcs : null;
    }

    // 39. FindTestCaseStartedBy(TestStepFinished)
    public TestCaseStarted? FindTestCaseStartedBy(TestStepFinished testStepFinished)
    {
        return _repository.TestCaseStartedById.TryGetValue(testStepFinished.TestCaseStartedId, out var tcs) ? tcs : null;
    }

    // 40. FindTestCaseFinishedBy(TestCaseStarted)
    public TestCaseFinished? FindTestCaseFinishedBy(TestCaseStarted testCaseStarted)
    {
        return _repository.TestCaseFinishedByTestCaseStartedId.TryGetValue(testCaseStarted.Id, out var finished) ? finished : null;
    }

    // 41. FindTestRunHookFinishedBy(TestRunHookStarted)
    public TestRunHookFinished? FindTestRunHookFinishedBy(TestRunHookStarted testRunHookStarted)
    {
        return _repository.TestRunHookFinishedByTestRunHookStartedId.TryGetValue(testRunHookStarted.Id, out var finished) ? finished : null;
    }

    // 42. FindTestRunHookStartedBy(TestRunHookFinished)
    public TestRunHookStarted? FindTestRunHookStartedBy(TestRunHookFinished testRunHookFinished)
    {
        return _repository.TestRunHookStartedById.TryGetValue(testRunHookFinished.TestRunHookStartedId, out var started) ? started : null;
    }

    // 43. FindTestRunDuration
    public System.TimeSpan? FindTestRunDuration()
    {
        if (_repository.TestRunStarted == null || _repository.TestRunFinished == null)
            return null;
        var start = Converters.ToDateTime(_repository.TestRunStarted.Timestamp);
        var finish = Converters.ToDateTime(_repository.TestRunFinished.Timestamp);
        return finish - start;
    }

    // 44. FindTestRunFinished
    public TestRunFinished? FindTestRunFinished() => _repository.TestRunFinished;

    // 45. FindTestRunStarted
    public TestRunStarted? FindTestRunStarted() => _repository.TestRunStarted;

    // 46. FindTestStepBy(TestStepStarted)
    public TestStep? FindTestStepBy(TestStepStarted testStepStarted)
    {
        return _repository.TestStepById.TryGetValue(testStepStarted.TestStepId, out var testStep) ? testStep : null;
    }

    // 47. FindTestStepBy(TestStepFinished)
    public TestStep? FindTestStepBy(TestStepFinished testStepFinished)
    {
        return _repository.TestStepById.TryGetValue(testStepFinished.TestStepId, out var testStep) ? testStep : null;
    }

    // 48. FindTestStepsStartedBy(TestCaseStarted)
    public IList<TestStepStarted> FindTestStepsStartedBy(TestCaseStarted testCaseStarted)
    {
        if (_repository.TestStepsStartedByTestCaseStartedId.TryGetValue(testCaseStarted.Id, out var steps))
        {
            return new List<TestStepStarted>(steps);
        }
        return new List<TestStepStarted>();
    }

    // 49. FindTestStepsStartedBy(TestCaseFinished)
    public IList<TestStepStarted> FindTestStepsStartedBy(TestCaseFinished testCaseFinished)
    {
        if (_repository.TestStepsStartedByTestCaseStartedId.TryGetValue(testCaseFinished.TestCaseStartedId, out var steps))
        {
            return new List<TestStepStarted>(steps);
        }
        return new List<TestStepStarted>();
    }

    // 50. FindTestStepsFinishedBy(TestCaseStarted)
    public IList<TestStepFinished> FindTestStepsFinishedBy(TestCaseStarted testCaseStarted)
    {
        if (_repository.TestStepsFinishedByTestCaseStartedId.TryGetValue(testCaseStarted.Id, out var steps))
        {
            return new List<TestStepFinished>(steps);
        }
        return new List<TestStepFinished>();
    }

    // 51. FindTestStepsFinishedBy(TestCaseFinished)
    public IList<TestStepFinished> FindTestStepsFinishedBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindTestStepsFinishedBy(testCaseStarted) : new List<TestStepFinished>();
    }

    // 52. FindTestStepFinishedAndTestStepBy(TestCaseStarted)
    public IList<(TestStepFinished, TestStep)> FindTestStepFinishedAndTestStepBy(TestCaseStarted testCaseStarted)
    {
        var finishedSteps = FindTestStepsFinishedBy(testCaseStarted);
        var result = new List<(TestStepFinished, TestStep)>();
        foreach (var testStepFinished in finishedSteps)
        {
            var testStep = FindTestStepBy(testStepFinished);
            if (testStep != null)
            {
                result.Add((testStepFinished, testStep));
            }
        }
        return result;
    }

    // FindLineageBy methods 
    public Lineage? FindLineageBy(GherkinDocument element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Feature element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Rule element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Scenario element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Examples element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(TableRow element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Pickle pickle)
    {
        var astNodeIds = pickle.AstNodeIds;
        var lastAstNodeId = astNodeIds.LastOrDefault();
        _repository.LineageById.TryGetValue(lastAstNodeId, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(TestCaseStarted testCaseStarted)
    {
        var pickle = FindPickleBy(testCaseStarted);
        if (pickle == null)
        {
            return null;
        }
        return FindLineageBy(pickle);
    }
}
